### PR TITLE
PUB-150: IngesterFormatter refactoring

### PR DESCRIPTION
**Pull request by basilisk487**

### CODE REVIEW

File: java-lib/src/main/java/com/wavefront/ingester/AbstractIngesterFormatter.java [hunk 1]
```diff
@@ -1,1000 +1,376 @@
...
-        satisfied = true;
-        element.consume(tokenQueue, point);
-      }
-
-      if (!satisfied) {
-        throw new RuntimeException("Expected at least one element, got none");
+    public void consume(StringParser parser, T target) {
+      String token = parser.next();
+      if (token == null)
+        throw new RuntimeException ("Missing value");
```
Comment (sushantdewan123) [inline java-lib/src/main/java/com/wavefront/ingester/AbstractIngesterFormatter.java]: extra space 
RuntimeException ("Missing value");

File: java-lib/src/main/java/com/wavefront/ingester/StringParser.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+  /**
+   * Retrieves the next available token, but does not advance the further, so multiple
+   * calls to peek() return the same value. The value is cached so performance
+   * penalty is negligible.
+   *
+   * @return next available token or null if end of line was reached
+   */
+  public String peek() {
+    if (input == null) throw new IllegalStateException("parse() must be called first");
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/StringParser.java]: I don't see a parse() call

File: java-lib/src/main/java/com/wavefront/ingester/StringParser.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+public class StringParser {
+  private int currentIndex = 0;
+  private final String input;
+  private String peek = null;
+
+  /**
+   * @param input string to parse at instance creation
+   */
+  public StringParser(String input) {
+    this.input = input;
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/StringParser.java]: just reject null inputs?

File: java-lib/src/main/java/com/wavefront/ingester/StringParser.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  /**
+   * Retrieves the next available token, but does not advance the further, so multiple
+   * calls to peek() return the same value. The value is cached so performance
+   * penalty is negligible.
+   *
+   * @return next available token or null if end of line was reached
+   */
+  public String peek() {
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/StringParser.java]: @Nullable

File: java-lib/src/main/java/com/wavefront/ingester/StringParser.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  public boolean hasNext() {
+    return peek() != null;
+  }
+
+  /**
+   * Retrieves the next available token and advances further.
+   *
+   * @return next available token or null if end of line was reached
+   */
+  public String next() {
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/StringParser.java]: @Nullable

File: java-lib/src/main/java/com/wavefront/ingester/StringParser.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    String token = peek();
+    peek = null;
+    return token;
+  }
+
+  private String advance() {
```
Comment (panghy) [inline java-lib/src/main/java/com/wavefront/ingester/StringParser.java]: @Nullable

File: java-lib/src/main/java/com/wavefront/ingester/StringParser.java (hunk 2 above)
Comment (basilisk487) [reply java-lib/src/main/java/com/wavefront/ingester/StringParser.java]: good catch, thanks! fixed. 
it was a carry-over from the first take where parser was threadlocal and re-usable. as it turned out, in jdk11 threadlocal carries about 9-10% performance penalty compared to just creating a new instance every time (creating a new parser is very cheap), so it didn't make much sense to keep it reusable.

File: java-lib/src/main/java/com/wavefront/ingester/StringParser.java (hunk 3 above)
Comment (basilisk487) [reply java-lib/src/main/java/com/wavefront/ingester/StringParser.java]: done

File: java-lib/src/main/java/com/wavefront/ingester/StringParser.java (hunk 5 above)
Comment (basilisk487) [reply java-lib/src/main/java/com/wavefront/ingester/StringParser.java]: done

File: java-lib/src/main/java/com/wavefront/ingester/StringParser.java (hunk 6 above)
Comment (basilisk487) [reply java-lib/src/main/java/com/wavefront/ingester/StringParser.java]: done

**Pull request merged**